relay-server/src/services/buffer/mod.rs

```diff
@@ -158,74 +165,88 @@ impl EnvelopeBufferService {
         Ok(())
     }
 
-    async fn handle_message(&mut self, message: EnvelopeBuffer) {
+    async fn handle_message(
+        &mut self,
+        buffer: &mut PolymorphicEnvelopeBuffer,
+        message: EnvelopeBuffer,
+    ) {
         match message {
             EnvelopeBuffer::Push(envelope) => {
                 // NOTE: This function assumes that a project state update for the relevant
                 // projects was already triggered (see XXX).
                 // For better separation of concerns, this prefetch should be triggered from here
                 // once buffer V1 has been removed.
                 relay_log::trace!("EnvelopeBufferService push");
-                self.push(envelope).await;
+                self.push(buffer, envelope).await;
             }
             EnvelopeBuffer::NotReady(project_key, envelope) => {
                 relay_log::trace!("EnvelopeBufferService project not ready");
-                self.buffer.mark_ready(&project_key, false);
+                buffer.mark_ready(&project_key, false);
                 relay_statsd::metric!(counter(RelayCounters::BufferEnvelopesReturned) += 1);
-                self.push(envelope).await;
+                self.push(buffer, envelope).await;
             }
             EnvelopeBuffer::Ready(project_key) => {
                 relay_log::trace!("EnvelopeBufferService project ready {}", &project_key);
-                self.buffer.mark_ready(&project_key, true);
+                buffer.mark_ready(&project_key, true);
             }
         };
         self.changes = true;
     }
 
-    async fn push(&mut self, envelope: Box<Envelope>) {
-        if let Err(e) = self.buffer.push(envelope).await {
+    async fn push(&mut self, buffer: &mut PolymorphicEnvelopeBuffer, envelope: Box<Envelope>) {
+        if let Err(e) = buffer.push(envelope).await {
             relay_log::error!(
                 error = &e as &dyn std::error::Error,
                 "failed to push envelope"
             );
         }
     }
 
-    fn update_observable_state(&self) {
+    fn update_observable_state(&self, buffer: &mut PolymorphicEnvelopeBuffer) {
         self.has_capacity
-            .store(self.buffer.has_capacity(), Ordering::Relaxed);
+            .store(buffer.has_capacity(), Ordering::Relaxed);
     }
 }
 
 impl Service for EnvelopeBufferService {
     type Interface = EnvelopeBuffer;
 
     fn spawn_handler(mut self, mut rx: Receiver<Self::Interface>) {
+        let config = self.config.clone();
+        let memory_checker = self.memory_checker.clone();
         tokio::spawn(async move {
+            let mut buffer = PolymorphicEnvelopeBuffer::from_config(&config, memory_checker)
+                .await
+                .expect("Envelope buffer couldn't be initialized from the config");
+            buffer.initialize().await;
```

Comment:
Any reason we don't call `initialize()` inside of `from_config()`? Is it the difference in error handling?

Reply:
Initialize is designed to be infallible but initially, I wanted to design the buffer with the initialization being independent of instantiation, because you don't need the initialization, this separation might be also useful for testing.